Description: libminkadaptor: add libcbor fallback for CBOR support
 The libminkadaptor library statically links libqcomtee, which defaults to
 QCBOR for CBOR encoding. However, it also supports fallback to libcbor
 when QCBOR is not available. This change ensures libminkadaptor
 correctly follows the same CBOR backend selection used by libqcomtee
 (i.e. where libqcomtee was compiler with libcbor instead of qcbor),
 enabling builds on distributions such as Debian and Fedora.

 Additionally, smcinvoke_client is updated to avoid direct QCBOR usage
 and to follow the same libcbor fallback mechanism. This improves build
 portability and ensures consistent CBOR handling across components.
Author: Abhinaba Rakshit <abhinaba.rakshit@oss.qualcomm.com>
Origin: upstream, https://github.com/qualcomm/minkipc/pull/58
Last-Update: 2026-04-30
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
Index: pkg-minkipc/libminkadaptor/CMakeLists.txt
===================================================================
--- pkg-minkipc.orig/libminkadaptor/CMakeLists.txt
+++ pkg-minkipc/libminkadaptor/CMakeLists.txt
@@ -35,7 +35,10 @@ set(QCBOR_DIR_HINT "" CACHE PATH "Hint p
 
 find_package(QCBOR REQUIRED)
 if(NOT QCBOR_FOUND)
-       message(FATAL_ERROR "QCBOR not found")
+	message(WARNING "QCBOR not found, falling back to libcbor")
+	set(USE_LIBCBOR TRUE)
+else()
+	set(USE_LIBCBOR FALSE)
 endif()
 
 # ''Source files''.
@@ -71,11 +74,19 @@ target_include_directories(minkadaptor
 	PRIVATE src
 )
 
-target_link_libraries(minkadaptor
-	PRIVATE ${QCOMTEE_LIBRARIES}
-	PRIVATE ${QCBOR_LIBRARIES}
-	PRIVATE ${CMAKE_THREAD_LIBS_INIT}
-)
+if (NOT USE_LIBCBOR)
+	target_link_libraries(minkadaptor
+		PRIVATE ${QCOMTEE_LIBRARIES}
+		PRIVATE ${QCBOR_LIBRARIES}
+		PRIVATE ${CMAKE_THREAD_LIBS_INIT}
+	)
+else()
+	target_link_libraries(minkadaptor
+		PRIVATE ${QCOMTEE_LIBRARIES}
+		PRIVATE cbor
+		PRIVATE ${CMAKE_THREAD_LIBS_INIT}
+	)
+endif()
 
 # ''Install targets''.
 
Index: pkg-minkipc/libminkadaptor/tests/smcinvoke_client/CMakeLists.txt
===================================================================
--- pkg-minkipc.orig/libminkadaptor/tests/smcinvoke_client/CMakeLists.txt
+++ pkg-minkipc/libminkadaptor/tests/smcinvoke_client/CMakeLists.txt
@@ -9,7 +9,15 @@ set(QCBOR_DIR_HINT "" CACHE PATH "Hint p
 
 find_package(QCBOR REQUIRED)
 if(NOT QCBOR_FOUND)
-	message(FATAL_ERROR "QCBOR not found")
+	message(WARNING "QCBOR not found, falling back to libcbor")
+	set(USE_LIBCBOR TRUE)
+else()
+	set(USE_LIBCBOR FALSE)
+endif()
+
+if(USE_LIBCBOR)
+	add_compile_definitions(${PROJECT_NAME}
+		PRIVATE -DUSE_LIBCBOR)
 endif()
 
 include_directories(${QCBOR_INCLUDE_DIRS})
@@ -84,9 +92,16 @@ target_include_directories(${PROJECT_NAM
 	PRIVATE idl
 )
 
-target_link_libraries(${PROJECT_NAME}
-	PRIVATE minkadaptor
-	PRIVATE ${QCBOR_LIBRARIES}
-)
+if(NOT USE_LIBCBOR)
+	target_link_libraries(${PROJECT_NAME}
+		PRIVATE minkadaptor
+		PRIVATE ${QCBOR_LIBRARIES}
+	)
+else()
+	target_link_libraries(${PROJECT_NAME}
+		PRIVATE minkadaptor
+		PRIVATE cbor
+	)
+endif()
 
 install(TARGETS ${PROJECT_NAME} DESTINATION "${CMAKE_INSTALL_BINDIR}")
Index: pkg-minkipc/libminkadaptor/tests/smcinvoke_client/src/smcinvoke_client.cpp
===================================================================
--- pkg-minkipc.orig/libminkadaptor/tests/smcinvoke_client/src/smcinvoke_client.cpp
+++ pkg-minkipc/libminkadaptor/tests/smcinvoke_client/src/smcinvoke_client.cpp
@@ -29,7 +29,12 @@
 #include "MinkCom.h"
 #include "tzecotestapp_uids.h"
 #include "tzt.h"
+#ifndef USE_LIBCBOR
 #include <qcbor/qcbor.h>
+#else
+#include <cbor.h>
+#include <string.h>
+#endif
 
 int32_t create_and_assign_mem_obj(Object, Object *);
 
@@ -61,6 +66,7 @@ static int64_t get_time_in_ms(void)
 	return (int64_t)(tv.tv_sec * 1000) + (int64_t)(tv.tv_usec / 1000);
 }
 
+#ifndef USE_LIBCBOR
 static int realloc_useful_buf(UsefulBuf *buf)
 {
 	void *ptr;
@@ -75,12 +81,29 @@ static int realloc_useful_buf(UsefulBuf
 
 	return 0;
 }
+#else
+static int alloc_cbor_mutable_data(cbor_mutable_data *buffer, size_t* buffer_size)
+{
+	void* ptr;
+
+	ptr = malloc(CREDENTIALS_BUF_SIZE_INC * sizeof(unsigned char));
+	if (!ptr)
+		return -1;
+
+	*buffer = (unsigned char*)ptr;
+	*buffer_size += CREDENTIALS_BUF_SIZE_INC;
+	memset(*buffer, 0, *buffer_size);
+
+	return 0;
+}
+#endif
 
 /* Get credentials. */
 static void* get_self_creds(size_t *buf_len)
 {
-	QCBOREncodeContext e_ctx;
 	void *credential_buf = nullptr;
+#ifndef USE_LIBCBOR
+	QCBOREncodeContext e_ctx;
 	UsefulBufC enc;
 	UsefulBuf creds_useful_buf = { NULL, 0 };
 
@@ -104,6 +127,49 @@ static void* get_self_creds(size_t *buf_
 	*buf_len = enc.len;
 
 	return credential_buf;
+#else
+	cbor_item_t *creds_map = NULL;
+	struct cbor_pair map_pair;
+	cbor_mutable_data buffer = NULL;
+	size_t buffer_size = 0;
+
+	if (alloc_cbor_mutable_data(&buffer, &buffer_size))
+		goto exit;
+
+	creds_map = cbor_new_definite_map(2);
+	if (creds_map == NULL)
+		goto exit;
+
+	map_pair.key = cbor_build_uint8(attr_uid);
+	map_pair.value = cbor_build_uint32((uint64_t)getuid());
+	if(!cbor_map_add(creds_map, map_pair)) {
+		cbor_decref(&map_pair.key);
+		cbor_decref(&map_pair.value);
+		goto exit;
+	}
+
+	map_pair.key = cbor_build_uint8(attr_system_time);
+	map_pair.value = cbor_build_uint64((uint64_t)get_time_in_ms());
+	if(!cbor_map_add(creds_map, map_pair)) {
+		cbor_decref(&map_pair.key);
+		cbor_decref(&map_pair.value);
+		goto exit;
+	}
+
+	buffer_size = cbor_serialize_map(creds_map, buffer, buffer_size);
+	if (buffer_size == 0)
+		goto exit;
+
+	/* Mark success of encoding */
+	credential_buf = (void *)buffer;
+	*buf_len = buffer_size;
+exit:
+	if(credential_buf == NULL && buffer)
+		free(buffer);
+	if(creds_map)
+		cbor_decref(&creds_map);
+	return credential_buf;
+#endif
 }
 
 int32_t create_and_assign_mem_obj(Object root, Object *argptr)
